### DISCUSSION TITLE
Several issues found in 3.0.7

### DISCUSSION THREAD

**Discussion by WMedDougJaquays** [nilsteampassnet/TeamPass#3713] [General]
I'm still trying to work through all of the errors we are receiving and issues we're having after upgrading to 3.x from 2.x (There's a lot.

Because there's several I didn't want to create an issue for each one, but will post them here to be reviewed and someone else can determine if my findings are worth being turned into issues.

1. I did create an issue for this, and then closed it when I found the problem and fixed it, but it should probably have a code update to actually fix it. #3700 - The addition of the "TP_USER" is incompatible with, at the very least, older installs. The column teampass_users.encrypted_psk was created with the "NOT NULL" setting, but the code to create the TP_USER attempts to insert a row with that column null... and does not error checking to make sure it works.
2. Several issues have been created about the use of the SECUREFILE constant in newer versions that is never defined. The way the application works, it actually has to be defined twice. Once in includes/config/include.php and once in include/config/settings.php. If you do not define it in both, you'll have issues. This is what I added to both: `> if (!defined('SECUREFILE')) {
>   define('SECUREFILE', 'teampass-seckey.txt');
> }`
3. The LDAP functions have the certificate check option hard-coded to LDAP_OPT_X_TLS_HARD, despite having a setting for the user to choose. This is true in both sources/identify.php:1227 and sources/users_queries.php:2545. This should be changed to user $SETTINGS['ldap_tls_certiface_check'].
4. In source/main.functions.php:1882, $attribute gets set to an empty string if $raison is null. at line 1896, $attribute[0] is referenced. This results in an error. $attribute should be set to Array('') if $raison is null to prevent the error.
5. There's an issue for our users (LDAP synched, if that matters. Haven't found the cause yet) where $_SESSION['user']['private_key'] is only populated the first time they login. For all subsequent logins, that key is set with an empty string. This, of course, breaks almost everything in the application.
6. We've also had an issue where logging in (Again, LDAP synched) sometimes results in an "invalid credentials" message despite the user's ability to login on the next attempt using the exactly same password pasted exactly the same way out of the clipboard. No error messages are logged that I can find, yet.

**Comment by WMedDougJaquays** [comment]
Fix for item 2 
```php
includes/config/include.php
72,74
> if (!defined('SECUREFILE')) {
>   define('SECUREFILE', 'teampass-seckey.txt');
> }
```

```php
includes/config/settings.php
26,28
> if (!defined('SECUREFILE')) {
>   define('SECUREFILE', 'teampass-seckey.txt');
> }
```

**Comment by WMedDougJaquays** [comment]
Fix for item 3 

```php
sources/identify.php
1227c1227
<             LDAP_OPT_X_TLS_REQUIRE_CERT => LDAP_OPT_X_TLS_HARD,
---
>             LDAP_OPT_X_TLS_REQUIRE_CERT => (isset($SETTINGS['ldap_tls_certiface_check']) ? $SETTINGS['ldap_tls_certiface_check'] : LDAP_OPT_X_TLS_HARD),
```
```php
sources/users.queries.php
2545c2545
<                     LDAP_OPT_X_TLS_REQUIRE_CERT => LDAP_OPT_X_TLS_HARD
---
>                     LDAP_OPT_X_TLS_REQUIRE_CERT => (isset($SETTINGS['ldap_tls_certiface_check']) ? $SETTINGS['ldap_tls_certiface_check'] : LDAP_OPT_X_TLS_HARD)
```

**Comment by WMedDougJaquays** [comment]
Fix for item 4
```php
sources/main.functions.php
1882c1882
<         $attribute = is_null($raison) === true ? '' : explode(' : ', $raison);
---
>         $attribute = is_null($raison) === true ? Array('') : explode(' : ', $raison);
```

**Comment by WMedDougJaquays** [comment]
I am at a loss for item 5. Everything is populated and looks accurate (from what I can tell) until it tries to decrypt the private_key stored in the DB.
sources/main.functions.php:2716 returns an empty value. I can't tell why.
```php
            return base64_encode((string) $cipher->decrypt(base64_decode($userPrivateKey)));
```
is the offending line of code. It is called by
sources/identify.php:1899
```php
        'private_key_clear' => decryptPrivateKey($passwordClear, $userInfo['private_key']),
```
which is called by
sources/identify.php:495
```php
        $returnKeys = prepareUserEncryptionKeys($userInfo, $passwordClear)
```
Both $userInfo and $passwordClear are populated and accurate when this function call executes.

I've followed it all the way down to includes/libraries/Encryption/phpseclib/Crypt/Base.php and everything looks fine until it tries to return the decrypted plain text, which is empty. My only guess is that there is a mismatch between how something is encrypted/hashed as stored and how the current library is trying to decrypt. I've generated new keys for the user I'm working with, but that changed nothing.

**Comment by Saibamen** [comment]
Should be fixed. Please close
